File: app/src/main/java/com/github/fo2rist/mclaren/ui/MainPresenter.java
package com.github.fo2rist.mclaren.ui;

import android.support.annotation.NonNull;

import com.github.fo2rist.mclaren.analytics.Events;
import com.github.fo2rist.mclaren.analytics.EventsLogger;
import com.github.fo2rist.mclaren.mvp.MainScreenContract;
import com.github.fo2rist.mclaren.repository.RaceCalendarRepository;
import com.github.fo2rist.mclaren.ui.models.RaceCalendar;
import javax.inject.Inject;

import static com.github.fo2rist.mclaren.utils.LinkUtils.getMcLarenCarLink;
import static com.github.fo2rist.mclaren.utils.LinkUtils.getMcLarenFormula1Link;

public class MainPresenter implements MainScreenContract.Presenter {
    private MainScreenContract.View view;
    private EventsLogger eventsLogger;
    private RaceCalendar raceCalendar;

    @Inject
    MainPresenter(EventsLogger eventsLogger, RaceCalendarRepository raceCalendarRepository) {
        this.eventsLogger = eventsLogger;
        this.raceCalendar = raceCalendarRepository.loadCurrentCalendar();
    }

    @Override
    public void onStart(@NonNull MainScreenContract.View view) {
        this.view = view;
        view.openStories();
        if (isRaceActive()) {
            view.showTransmissionButton();
        }
    }

    @Override
    public void onRestart(@NonNull MainScreenContract.View view) {
        this.view = view;
    }

    private boolean isRaceActive() {
        return raceCalendar.getActiveEvent() != null;
    }

    @Override
    public void onStoriesClicked() {
        view.openStories();
        eventsLogger.logViewEvent(Events.MENU_STORIES);
    }

    @Override
    public void onCircuitsClicked() {
        view.openCircuits();
        eventsLogger.logViewEvent(Events.MENU_CIRCUITS);
    }

    @Override
    public void onDriversClicked() {
        view.openDrivers();
        eventsLogger.logViewEvent(Events.MENU_DRIVERS);
    }

    @Override
    public void onCarClicked() {
        view.navigateTo(
                getMcLarenCarLink());

        eventsLogger.logViewEvent(Events.MENU_CAR);
    }

    @Override
    public void onOfficialSiteClicked() {
        view.navigateTo(
                getMcLarenFormula1Link());
        eventsLogger.logViewEvent(Events.MENU_SITE);
    }

    @Override
    public void onAboutClicked() {
        view.openAboutScreen();
        eventsLogger.logViewEvent(Events.MENU_ABOUT);
    }

    @Override
    public void onTransmissionCenterClicked() {
        view.openTransmissionCenter();
        eventsLogger.logViewEvent(Events.MENU_TRANSMISSION_CENTER);
    }
}

File: app/src/main/java/com/github/fo2rist/mclaren/ui/MainPresenter.kt
package com.github.fo2rist.mclaren.ui

import com.github.fo2rist.mclaren.analytics.Events
import com.github.fo2rist.mclaren.analytics.EventsLogger
import com.github.fo2rist.mclaren.mvp.MainScreenContract
import com.github.fo2rist.mclaren.repository.remoteconfig.RaceCalendarRepository
import com.github.fo2rist.mclaren.ui.models.RaceCalendar
import com.github.fo2rist.mclaren.utils.LinkUtils.getMcLarenCarLink
import com.github.fo2rist.mclaren.utils.LinkUtils.getMcLarenFormula1Link
import javax.inject.Inject

/**
 * Presenter for [MainActivity].
 * Controls high-level navigation to other components of the screen.
 */
class MainPresenter @Inject constructor(
    override var view: MainScreenContract.View,
    private val eventsLogger: EventsLogger,
    private val raceCalendarRepository: RaceCalendarRepository
) : MainScreenContract.Presenter {

    private val raceCalendar: RaceCalendar by lazy { raceCalendarRepository.loadCalendar() }

    override fun onStart() {
        openStories()
    }

    override fun onStoriesClicked() {
        openStories()
    }

    private fun openStories() {
        view.openStories()
        eventsLogger.overrideScreenName(Events.Screen.MENU_STORIES)

        val activeEvent = raceCalendar.getActiveEvent()
        val upcomingEvent = raceCalendar.getNextEvent()
        if (activeEvent != null) {
            // can notify users about the race in progress
            // via view.showTransmissionButton()
            // do nothing for now (used to launch transmission screen)
        } else if (upcomingEvent != null) {
            view.showUpcomingEventButton(upcomingEvent.grandPrixName, upcomingEvent.practice1DateTime!!)
        }

    }

    override fun onTeamTwitterClicked() {
        view.openTweets()
        eventsLogger.overrideScreenName(Events.Screen.MENU_TEAM_TWITTER)

        view.hideFloatingButtons();
    }

    override fun onSeasonCalendarClicked() {
        view.openCircuits()
        eventsLogger.overrideScreenName(Events.Screen.MENU_CALENDAR)

        view.hideFloatingButtons();
    }

    override fun onDriversClicked() {
        view.openDrivers()
        eventsLogger.overrideScreenName(Events.Screen.MENU_DRIVERS)

        view.hideFloatingButtons();
    }

    override fun onCarClicked() {
        val mcLarenCarPageLink = getMcLarenCarLink()
        view.navigateTo(mcLarenCarPageLink)
        eventsLogger.logExternalNavigation(Events.Screen.MENU_CAR, mcLarenCarPageLink)
    }

    override fun onOfficialSiteClicked() {
        val mcLarenF1PageLink = getMcLarenFormula1Link()
        view.navigateTo(mcLarenF1PageLink)
        eventsLogger.logExternalNavigation(Events.Screen.MENU_SITE, mcLarenF1PageLink)
    }

    override fun onAboutClicked() {
        view.navigateToAboutScreen()
    }

    override fun onTransmissionCenterClicked() {
        view.openTransmissionCenter()
        eventsLogger.overrideScreenName(Events.Screen.TRANSMISSION_CENTER)
    }

    override fun onUpcomingEventClicked() {
        val eventToOpen = raceCalendar.getActiveEvent()
                ?: raceCalendar.getNextEvent()
                ?: return
        view.navigateToCircuitScreen(raceCalendar.indexOf(eventToOpen)) //the index is never out of bounds here
    }
}
